python/v9.py

```diff
@@ -0,0 +1,89 @@
+import json
+from urllib.parse import parse_qs, quote, unquote
+
+CRIT_ERROR_STRING = quote("{\"response_body\":\"\",\"http_response_code\":543,\"internal_error\":true,\"error_message\":\"can't even serialize what we're working with!\"}") + "\n"
+
+
+def serialize_response(response_code, body, *, internal_error=None):
+    try:
+        resp = {
+            "response_body": str(body),
+            "http_response_code": int(response_code)
+        }
+        if internal_error is not None:
+            resp["error_message"] = str(internal_error)
+
+        return str.encode(quote(json.dumps(resp)) + "\n")
+    except Exception:
+        return str.encode(CRIT_ERROR_STRING)
+
+
+class V9Component(object):
+    def __init__(self, in_file, out_file):
+        self.in_file = in_file
+        self.out_file = out_file
+        self.functions = {}
+
+    def register_operation(self, name, function):
+        self.functions[name] = function
+
+    def _get_response(self, line):
+        try:
+            deserialized_line = json.loads(unquote(line.decode("utf-8")))
+        except Exception as e:
+            # 543 = internal error
+            # (This means the node has messed up serializing the line)
+            return serialize_response(543, "", internal_error="Invalid line " + str(line) + " exception " + repr(e))
+
+        try:
+            function = deserialized_line["called_function"]
+            http_method = deserialized_line["http_method"]
+            path = deserialized_line["path"]
+            request_arguments = parse_qs(deserialized_line["request_arguments"])
+            request_body = deserialized_line["request_body"]
+
+            if function not in self.functions:
+                return serialize_response(404, "Function not found!")
+
+            try:
+                (response_code, response_body) = self.functions[function](http_method, path, request_arguments, request_body)
+            except Exception as e:
+                return serialize_response(500, "User function failed " + repr(e))
+
+            return serialize_response(response_code, response_body)
+        except KeyError:
+            # 543 = internal error
+            # (This means the node has messed up serializing the line)
+            return serialize_response(543, "", internal_error="Invalid request " + deserialized_line)
+
+    def loop(self):
```

Comment:
Is there anyway to make this less nested, because it's pretty confusing

Reply:
There really isn't much actual code here. I probably could remove the inner try/except. That's still 4 levels of depth though... What part here do you find confusing specifically?

Reply:
Mostly just complaining about the nested indents. What happens if the loop starts but the pipe is already empty?

Reply:
It doesn't matter if there is stuff in the pipe or not -- either we block waiting on read, we fail a read, or we read a line. None of those three cases is handled erroneously as far as I can see?

Reply:
Got it, the block happens in read, I didn't realize that.